In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos=pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
correction={'dateCrawled':'date_crawled',
            'name':'name',
            'seller':'seller',
            'offerType':'offer_type',
            'price':'price',
            'abtest':'abtest',
            'vehicleType':'vehicle_type',
            'yearOfRegistration':'registration_year',
            'gearbox':'gearbox',
            'powerPS':'powerPS',
            'model':'model',
            'odometer':'odometer',
            'monthOfRegistration':'registration_month',
            'fuelType':'fuel_type',
            'brand':'brand',
            'notRepairedDamage':'unrepaired_damage',
            'dateCreated':'ad_created',
            'nrOfPictures':'nr_of_pictures',
            'postalCode':'postal_code',
            'lastSeen':'last_seen'}
            
autos.rename(columns={'dateCrawled':'date_crawled',
            'name':'name',
            'seller':'seller',
            'offerType':'offer_type',
            'price':'price',
            'abtest':'abtest',
            'vehicleType':'vehicle_type',
            'yearOfRegistration':'registration_year',
            'gearbox':'gearbox',
            'powerPS':'powerPS',
            'model':'model',
            'odometer':'odometer',
            'monthOfRegistration':'registration_month',
            'fuelType':'fuel_type',
            'brand':'brand',
            'notRepairedDamage':'unrepaired_damage',
            'dateCreated':'ad_created',
            'nrOfPictures':'nr_of_pictures',
            'postalCode':'postal_code',
            'lastSeen':'last_seen'}, inplace=True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [9]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [10]:
autos.describe()

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [11]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
5         $7,900
6           $300
7         $1,990
8           $250
9           $590
10          $999
11          $350
12        $5,299
13        $1,350
14        $3,999
15       $18,900
16          $350
17        $5,500
18          $300
19        $4,150
20        $3,500
21       $41,500
22       $25,450
23        $7,999
24       $48,500
25           $90
26          $777
27            $0
28        $5,250
29        $4,999
          ...   
49970    $15,800
49971       $950
49972     $3,300
49973     $6,000
49974         $0
49975     $9,700
49976     $5,900
49977     $5,500
49978       $900
49979    $11,000
49980       $400
49981     $2,000
49982     $1,950
49983       $600
49984         $0
49985     $1,000
49986    $15,900
49987    $21,990
49988     $9,550
49989       $150
49990    $17,500
49991       $500
49992     $4,800
49993     $1,650
49994     $5,000
49995    $24,900
49996     $1,980
49997    $13,2

In [12]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int) 

In [13]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [14]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
5        150,000km
6        150,000km
7        150,000km
8        150,000km
9        150,000km
10       150,000km
11       150,000km
12        50,000km
13       150,000km
14       150,000km
15        80,000km
16       150,000km
17       150,000km
18       150,000km
19       150,000km
20       150,000km
21       150,000km
22        10,000km
23       150,000km
24        30,000km
25       150,000km
26       125,000km
27       150,000km
28       150,000km
29       150,000km
           ...    
49970     60,000km
49971    150,000km
49972    150,000km
49973    150,000km
49974    150,000km
49975    100,000km
49976    150,000km
49977    150,000km
49978    150,000km
49979     70,000km
49980    125,000km
49981    150,000km
49982     90,000km
49983    150,000km
49984    150,000km
49985    150,000km
49986    125,000km
49987     50,000km
49988    150,000km
49989    150,000km
49990     30,000km
49991    150

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [16]:
autos['odometer']

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [17]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [18]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [19]:
autos['odometer_km'].max()

150000

In [20]:
autos['odometer_km'].min()

5000

In [21]:
autos['price'].max()

99999999

In [22]:
autos['price'].min()

0

In [23]:
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [24]:
autos.shape

(50000, 20)

In [25]:
autos['odometer_km'].unique().shape

(13,)

In [26]:
autos['price'].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [27]:
autos['price'].unique().shape

(2357,)

In [28]:
autos['odometer_km'].mean()

125732.7

In [29]:
autos['price'].mean()

9840.04376

In [30]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [31]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [32]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox       powerPS  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [33]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [34]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [35]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [36]:
autos['price'].value_counts()

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

In [37]:
autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [38]:
autos['price']= autos.loc[autos['price'].between(2,350000),'price']

In [39]:
autos['price'].value_counts().sort_index(ascending=True)

2.0          3
3.0          1
5.0          2
8.0          1
9.0          1
10.0         7
11.0         2
12.0         3
13.0         2
14.0         1
15.0         2
17.0         3
18.0         1
20.0         4
25.0         5
29.0         1
30.0         7
35.0         1
40.0         6
45.0         4
47.0         1
49.0         4
50.0        49
55.0         2
59.0         1
60.0         9
65.0         5
66.0         1
70.0        10
75.0         5
            ..
119500.0     1
119900.0     1
120000.0     2
128000.0     1
129000.0     1
130000.0     1
135000.0     1
137999.0     1
139997.0     1
145000.0     1
151990.0     1
155000.0     1
163500.0     1
163991.0     1
169000.0     1
169999.0     1
175000.0     1
180000.0     1
190000.0     1
194000.0     1
197000.0     1
198000.0     1
220000.0     1
250000.0     1
259000.0     1
265000.0     1
295000.0     1
299000.0     1
345000.0     1
350000.0     1
Name: price, Length: 2345, dtype: int64

In [40]:
autos['price'].describe()

count     48409.000000
mean       5907.909707
std        9068.263463
min           2.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

outliers atilmadan onceki rakamlar:

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [41]:
autos['price'].describe(include='all')

count     48409.000000
mean       5907.909707
std        9068.263463
min           2.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [42]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [43]:
autos[['registration_year','registration_month']][0:5]

registration_year  registration_month
0               2004                   3
1               1997                   6
2               2009                   7
3               2007                   6
4               2003                   7

In [44]:
autos[['date_crawled','ad_created','last_seen']].describe()

date_crawled           ad_created            last_seen
count                 50000                50000                50000
unique                48213                   76                39481
top     2016-03-09 11:54:38  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1946                    8

In [45]:
autos[['registration_year','registration_month']].describe()

registration_year  registration_month
count       50000.000000        50000.000000
mean         2005.073280            5.723360
std           105.712813            3.711984
min          1000.000000            0.000000
25%          1999.000000            3.000000
50%          2003.000000            6.000000
75%          2008.000000            9.000000
max          9999.000000           12.000000

In [46]:
autos['date_crawled'].value_counts()

2016-03-09 11:54:38    3
2016-03-12 16:06:22    3
2016-03-23 19:38:20    3
2016-03-27 22:55:05    3
2016-03-29 23:42:13    3
2016-03-30 17:37:35    3
2016-03-10 15:36:24    3
2016-04-02 11:37:04    3
2016-03-22 09:51:06    3
2016-03-05 16:57:05    3
2016-03-25 19:57:10    3
2016-03-19 17:36:18    3
2016-04-02 15:49:30    3
2016-03-08 10:40:35    3
2016-03-11 22:38:16    3
2016-03-16 21:50:53    3
2016-03-14 20:50:02    3
2016-03-30 19:48:02    3
2016-04-04 16:40:33    3
2016-03-23 18:39:34    3
2016-03-21 16:37:21    3
2016-03-21 20:37:19    3
2016-03-14 19:47:50    2
2016-03-29 13:38:54    2
2016-03-22 18:48:39    2
2016-03-30 02:57:54    2
2016-03-11 10:55:25    2
2016-03-08 19:40:38    2
2016-03-10 12:37:57    2
2016-03-16 21:50:16    2
                      ..
2016-03-20 13:58:00    1
2016-03-07 17:39:32    1
2016-03-15 18:54:51    1
2016-03-16 19:47:30    1
2016-03-29 20:46:53    1
2016-03-14 09:50:53    1
2016-03-21 18:48:02    1
2016-04-04 19:42:54    1
2016-03-22 20:56:07    1


In [47]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
27       2016-03-27
28       2016-03-19
29       2016-04-02
            ...    
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49974    2016-03-20
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49984    2016-03-31
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


In [48]:
autos['date_crawled'].str[:10].value_counts().sort_index(ascending=False)

2016-04-07      71
2016-04-06     159
2016-04-05     655
2016-04-04    1826
2016-04-03    1934
2016-04-02    1770
2016-04-01    1690
2016-03-31    1596
2016-03-30    1681
2016-03-29    1709
2016-03-28    1742
2016-03-27    1552
2016-03-26    1624
2016-03-25    1587
2016-03-24    1455
2016-03-23    1619
2016-03-22    1647
2016-03-21    1876
2016-03-20    1891
2016-03-19    1745
2016-03-18     653
2016-03-17    1576
2016-03-16    1475
2016-03-15    1699
2016-03-14    1831
2016-03-13     778
2016-03-12    1839
2016-03-11    1624
2016-03-10    1606
2016-03-09    1661
2016-03-08    1665
2016-03-07    1798
2016-03-06     697
2016-03-05    1269
Name: date_crawled, dtype: int64

In [49]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False)

2016-04-03    0.03868
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-12    0.03678
2016-03-14    0.03662
2016-04-04    0.03652
2016-03-07    0.03596
2016-04-02    0.03540
2016-03-19    0.03490
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-15    0.03398
2016-04-01    0.03380
2016-03-30    0.03362
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-22    0.03294
2016-03-11    0.03248
2016-03-26    0.03248
2016-03-23    0.03238
2016-03-10    0.03212
2016-03-31    0.03192
2016-03-25    0.03174
2016-03-17    0.03152
2016-03-27    0.03104
2016-03-16    0.02950
2016-03-24    0.02910
2016-03-05    0.02538
2016-03-13    0.01556
2016-03-06    0.01394
2016-04-05    0.01310
2016-03-18    0.01306
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [50]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [51]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index().max()

0.03868

date range is 2016-03-05 and 2016-04-07
the date of max observation is 2016-04-03

In [52]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
               ...   
2016-03-09    0.03324
2016-03-10    0.03186
2016-03-11    0.03278
2016-03-12    0.03662
2016-03-13    0.01692
2016-03-14    0.03522
2016-03-15    0.03374
2016-03-16    0.03000
2016-03-17    0.03120
2016-03-18    0.01372
2016-03-19    0.03384
2016-03-20    0.03786
2016-03-21    0.03772
2016-03-22    0.03280
2016-03-23

In [53]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index().max()

0.03892

date range is 2015-06-11 and 2016-04-07
the date of max observation is 2016-04-03. it is the same date of 'date crawled' column. what happened on that day??

In [54]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [55]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index().max()

0.221

date range is 2016-03-05 and 2016-04-07
the date of max observation is 2016-04-06
%22 of the total observation has the date of 2016-04-06. why?? it is an outlier.

In [56]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1000    0.00002
1001    0.00002
1111    0.00002
1500    0.00002
1800    0.00004
1910    0.00018
1927    0.00002
1929    0.00002
1931    0.00002
1934    0.00004
1937    0.00008
1938    0.00002
1939    0.00002
1941    0.00004
1943    0.00002
1948    0.00002
1950    0.00006
1951    0.00004
1952    0.00002
1953    0.00002
1954    0.00004
1955    0.00004
1956    0.00010
1957    0.00004
1958    0.00008
1959    0.00014
1960    0.00068
1961    0.00012
1962    0.00008
1963    0.00018
         ...   
2001    0.05406
2002    0.05066
2003    0.05454
2004    0.05474
2005    0.06030
2006    0.05416
2007    0.04608
2008    0.04462
2009    0.04196
2010    0.03194
2011    0.03268
2012    0.02646
2013    0.01612
2014    0.01332
2015    0.00798
2016    0.02632
2017    0.02906
2018    0.00984
2019    0.00006
2800    0.00002
4100    0.00002
4500    0.00002
4800    0.00002
5000    0.00008
5911    0.00002
6200    0.00002
8888    0.00002
9000    0.00004
9996    0.00002
9999    0.00008
Name: registration_year,

In [57]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index().max()

0.06708

In [58]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

there are some inaccurate years (before 1910 and after 2016) in the data, we should exlude them. 
Lets take the data between these 2 dates. why 2016? Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.

In [59]:
autos['registration_year']=autos.loc[autos['registration_year'].between(1910,2019),'registration_year']

In [60]:
autos['registration_year']

0        2004.0
1        1997.0
2        2009.0
3        2007.0
4        2003.0
5        2006.0
6        1995.0
7        1998.0
8        2000.0
9        1997.0
10       2017.0
11       2000.0
12       2010.0
13       1999.0
14       2007.0
15       1982.0
16       1999.0
17       1990.0
18       1995.0
19       2004.0
20       2003.0
21       2004.0
22       2015.0
23       2010.0
24       2014.0
25       1996.0
26       1992.0
27       2005.0
28       2007.0
29       2004.0
          ...  
49970    2010.0
49971    2001.0
49972    2004.0
49973    2004.0
49974    1983.0
49975    2012.0
49976    1992.0
49977    2003.0
49978    1996.0
49979    2011.0
49980    1995.0
49981    1998.0
49982    2004.0
49983    1999.0
49984    2000.0
49985    1995.0
49986    2010.0
49987    2013.0
49988    2001.0
49989    1997.0
49990    2012.0
49991    2016.0
49992    2009.0
49993    1997.0
49994    2001.0
49995    2011.0
49996    1996.0
49997    2014.0
49998    2013.0
49999    1996.0
Name: registration_year,

In [61]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

 1910.0    0.00018
 1927.0    0.00002
 1929.0    0.00002
 1931.0    0.00002
 1934.0    0.00004
 1937.0    0.00008
 1938.0    0.00002
 1939.0    0.00002
 1941.0    0.00004
 1943.0    0.00002
 1948.0    0.00002
 1950.0    0.00006
 1951.0    0.00004
 1952.0    0.00002
 1953.0    0.00002
 1954.0    0.00004
 1955.0    0.00004
 1956.0    0.00010
 1957.0    0.00004
 1958.0    0.00008
 1959.0    0.00014
 1960.0    0.00068
 1961.0    0.00012
 1962.0    0.00008
 1963.0    0.00018
 1964.0    0.00024
 1965.0    0.00034
 1966.0    0.00044
 1967.0    0.00054
 1968.0    0.00052
            ...   
 1991.0    0.00712
 1992.0    0.00782
 1993.0    0.00890
 1994.0    0.01320
 1995.0    0.02626
 1996.0    0.02888
 1997.0    0.04056
 1998.0    0.04906
 1999.0    0.06000
 2000.0    0.06708
 2001.0    0.05406
 2002.0    0.05066
 2003.0    0.05454
 2004.0    0.05474
 2005.0    0.06030
 2006.0    0.05416
 2007.0    0.04608
 2008.0    0.04462
 2009.0    0.04196
 2010.0    0.03194
 2011.0    0.03268
 2012.0    0

In [62]:
autos['registration_year'].describe()

count    49976.000000
mean      2003.366836
std          7.693943
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

decribe before exluding the outliers:

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [63]:
autos['brand']

0               peugeot
1                   bmw
2            volkswagen
3                 smart
4                  ford
5              chrysler
6            volkswagen
7            volkswagen
8                  seat
9               renault
10           volkswagen
11        mercedes_benz
12                smart
13                 audi
14              renault
15       sonstige_autos
16                 opel
17           volkswagen
18                  bmw
19                mazda
20                 audi
21              porsche
22                 mini
23              peugeot
24                  bmw
25                 ford
26           volkswagen
27                 ford
28                 mini
29        mercedes_benz
              ...      
49970           citroen
49971        volkswagen
49972     mercedes_benz
49973     mercedes_benz
49974        volkswagen
49975             honda
49976              audi
49977     mercedes_benz
49978     mercedes_benz
49979        volkswagen
49980           

In [64]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [65]:
sorted_brand=autos['brand'].value_counts(normalize=True).sort_values(ascending=False)

In [66]:
sorted_brand

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

In [67]:
sorted_brand[sorted_brand>0.05]

volkswagen       0.21374
opel             0.10922
bmw              0.10858
mercedes_benz    0.09468
audi             0.08566
ford             0.06958
Name: brand, dtype: float64

In [68]:
autos['brand'].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
citroen             701
smart               701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [69]:
autos['brand'].value_counts().index

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'daewoo', 'trabant', 'jaguar', 'rover', 'lancia',
       'lada'],
      dtype='object')

In [70]:
aggregate_label=sorted_brand[sorted_brand>0.05].index

In [71]:
aggregate_label

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [72]:
aggregate_data={}
for i in aggregate_label:
    mean_price=autos.loc[autos['brand']==i,'price'].mean()
    aggregate_data[i]=mean_price
    

In [73]:
aggregate_data

{'audi': 9241.752587244284,
 'bmw': 8294.405101846563,
 'ford': 3736.143111111111,
 'mercedes_benz': 8565.483606557376,
 'opel': 2959.746095238095,
 'volkswagen': 5346.960516103997}

In [74]:
brand_mean_prices_series=pd.Series(aggregate_data)

In [75]:
brand_mean_prices_series

audi             9241.752587
bmw              8294.405102
ford             3736.143111
mercedes_benz    8565.483607
opel             2959.746095
volkswagen       5346.960516
dtype: float64

In [76]:
brand_mean_prices_series.sort_values(ascending=False)

audi             9241.752587
mercedes_benz    8565.483607
bmw              8294.405102
volkswagen       5346.960516
ford             3736.143111
opel             2959.746095
dtype: float64

Audi, BMW and Mercedes Benz are more expensive Ford and Opel are less expensive Volkswagen is in between

In [77]:
df = pd.DataFrame(brand_mean_prices_series, columns=['mean_price'])

In [78]:
df

mean_price
audi           9241.752587
bmw            8294.405102
ford           3736.143111
mercedes_benz  8565.483607
opel           2959.746095
volkswagen     5346.960516

In [79]:
aggregate_data2={}
for i in aggregate_label:
    mean_mileage=autos.loc[autos['brand']==i,'odometer_km'].mean()
    aggregate_data2[i]=mean_mileage
    

In [80]:
aggregate_data2

{'audi': 129643.9411627364,
 'bmw': 132521.64302818198,
 'ford': 124131.93446392642,
 'mercedes_benz': 130886.14279678918,
 'opel': 129298.66324848929,
 'volkswagen': 128955.27276129878}

In [81]:
aggregate_data

{'audi': 9241.752587244284,
 'bmw': 8294.405101846563,
 'ford': 3736.143111111111,
 'mercedes_benz': 8565.483606557376,
 'opel': 2959.746095238095,
 'volkswagen': 5346.960516103997}

In [82]:
mean_prices_series=pd.Series(aggregate_data)

In [83]:
mean_prices_series

audi             9241.752587
bmw              8294.405102
ford             3736.143111
mercedes_benz    8565.483607
opel             2959.746095
volkswagen       5346.960516
dtype: float64

In [84]:
mean_mileage_series=pd.Series(aggregate_data2)

In [85]:
mean_mileage_series

audi             129643.941163
bmw              132521.643028
ford             124131.934464
mercedes_benz    130886.142797
opel             129298.663248
volkswagen       128955.272761
dtype: float64

In [86]:
df = pd.DataFrame(mean_prices_series, columns=['mean_price'])

In [87]:
df

mean_price
audi           9241.752587
bmw            8294.405102
ford           3736.143111
mercedes_benz  8565.483607
opel           2959.746095
volkswagen     5346.960516

In [88]:
df['mean_mileage']=mean_mileage_series

In [89]:
df

mean_price   mean_mileage
audi           9241.752587  129643.941163
bmw            8294.405102  132521.643028
ford           3736.143111  124131.934464
mercedes_benz  8565.483607  130886.142797
opel           2959.746095  129298.663248
volkswagen     5346.960516  128955.272761

In [94]:
df.sort_values('mean_price', ascending=False, inplace=True)

In [93]:
df

mean_price   mean_mileage
audi           9241.752587  129643.941163
mercedes_benz  8565.483607  130886.142797
bmw            8294.405102  132521.643028
volkswagen     5346.960516  128955.272761
ford           3736.143111  124131.934464
opel           2959.746095  129298.663248

In [96]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 48409 non-null float64
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     49976 non-null float64
gearbox               47320 non-null object
powerPS               50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: float6

In [98]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus             2004.0   
1  privat    Angebot  8500.0  control    limousine             1997.0   
2  privat    Angebot  8990.0     test    limousine             2009.0   
3  privat    Angebot  4350.0  control   kleinwagen             2007.0   
4  privat    Angebot  1350.0     test        kombi             2003.0   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [101]:
correction={'manuell':'manuel', 'automatik':'automatic'}
autos['gearbox']=autos['gearbox'].map(correction)

In [102]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus             2004.0   
1  privat    Angebot  8500.0  control    limousine             1997.0   
2  privat    Angebot  8990.0     test    limousine             2009.0   
3  privat    Angebot  4350.0  control   kleinwagen             2007.0   
4  privat    Angebot  1350.0     test        kombi             2003.0   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0     manuel      158  andere       150000                   3       lpg   
1  automatic      286     7er       150000                   6    benzin   
2     manuel      102    golf        70000                   7    benzin   
3  automatic       71  fortwo        70000                   6    benzin   
4     manuel        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [108]:
list=['date_crawled','ad_created','last_seen']
for i in list:
    autos[i]=autos[i].str[:10]

In [109]:
autos.head()

date_crawled                                               name  seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2   2016-03-26                         Volkswagen_Golf_1.6_United  privat   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus             2004.0     manuel   
1    Angebot  8500.0  control    limousine             1997.0  automatic   
2    Angebot  8990.0     test    limousine             2009.0     manuel   
3    Angebot  4350.0  control   kleinwagen             2007.0  automatic   
4    Angebot  1350.0     test        kombi             2003.0     manuel   

   powerPS   model  odometer_km  registration_month fuel_type       brand  \
0      158  andere       150000                   3       lpg     peugeot   
1      286     7er       150000                   6    benzin         bmw   
2      102    golf        70000                   7    benzin  volkswagen   
3       71  fortwo        70000                   6    benzin       smart   
4        0   focus       150000                   7    benzin        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
0              nein  2016-03-26               0        79588  2016-04-06  
1              nein  2016-04-04               0        71034  2016-04-06  
2              nein  2016-03-26               0        35394  2016-04-06  
3              nein  2016-03-12               0        33729  2016-03-15  
4              nein  2016-04-01               0        39218  2016-04-01

In [116]:
list=['date_crawled','ad_created','last_seen']
for i in list:
    autos[i]=autos[i].str.replace('-','').astype(int)


In [117]:
autos.head()

date_crawled                                               name  seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2      20160326                         Volkswagen_Golf_1.6_United  privat   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus             2004.0     manuel   
1    Angebot  8500.0  control    limousine             1997.0  automatic   
2    Angebot  8990.0     test    limousine             2009.0     manuel   
3    Angebot  4350.0  control   kleinwagen             2007.0  automatic   
4    Angebot  1350.0     test        kombi             2003.0     manuel   

   powerPS   model  odometer_km  registration_month fuel_type       brand  \
0      158  andere       150000                   3       lpg     peugeot   
1      286     7er       150000                   6    benzin         bmw   
2      102    golf        70000                   7    benzin  volkswagen   
3       71  fortwo        70000                   6    benzin       smart   
4        0   focus       150000                   7    benzin        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code  last_seen  
0              nein    20160326               0        79588   20160406  
1              nein    20160404               0        71034   20160406  
2              nein    20160326               0        35394   20160406  
3              nein    20160312               0        33729   20160315  
4              nein    20160401               0        39218   20160401

In [118]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null int64
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 48409 non-null float64
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     49976 non-null float64
gearbox               47320 non-null object
powerPS               50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null int64
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null int64
dtypes: float64(2

In [121]:
autos['brand-model']=autos['brand']+autos['model']

In [122]:
autos.head()

date_crawled                                               name  seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2      20160326                         Volkswagen_Golf_1.6_United  privat   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus             2004.0     manuel   
1    Angebot  8500.0  control    limousine             1997.0  automatic   
2    Angebot  8990.0     test    limousine             2009.0     manuel   
3    Angebot  4350.0  control   kleinwagen             2007.0  automatic   
4    Angebot  1350.0     test        kombi             2003.0     manuel   

   powerPS       ...       odometer_km  registration_month  fuel_type  \
0      158       ...            150000                   3        lpg   
1      286       ...            150000                   6     benzin   
2      102       ...             70000                   7     benzin   
3       71       ...             70000                   6     benzin   
4        0       ...            150000                   7     benzin   

        brand unrepaired_damage ad_created  nr_of_pictures  postal_code  \
0     peugeot              nein   20160326               0        79588   
1         bmw              nein   20160404               0        71034   
2  volkswagen              nein   20160326               0        35394   
3       smart              nein   20160312               0        33729   
4        ford              nein   20160401               0        39218   

   last_seen     brand-model  
0   20160406   peugeotandere  
1   20160406          bmw7er  
2   20160406  volkswagengolf  
3   20160315     smartfortwo  
4   20160401       fordfocus  

[5 rows x 21 columns]

In [125]:
autos['brand-model'].value_counts()

volkswagengolf                  4024
bmw3er                          2761
volkswagenpolo                  1757
opelcorsa                       1735
opelastra                       1454
volkswagenpassat                1425
audia4                          1291
bmw5er                          1183
mercedes_benzc_klasse           1172
mercedes_benze_klasse           1001
audia3                           882
audia6                           835
fordfocus                        811
fordfiesta                       769
volkswagentransporter            704
renaulttwingo                    677
peugeot2_reihe                   621
mercedes_benza_klasse            588
opelvectra                       574
smartfortwo                      574
bmw1er                           538
renaultclio                      509
fordmondeo                       496
volkswagentouran                 468
fiatpunto                        458
mercedes_benzandere              446
opelzafira                       423
f

In [146]:
list2=autos['odometer_km'].unique()

In [147]:
list2

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [149]:
mileage_price={}
for i in list2:
    mean_price=autos.loc[autos['odometer_km']==i,'price'].mean()
    mileage_price[i]=mean_price

In [150]:
mileage_price

{5000: 8750.483435582822,
 10000: 20491.216,
 20000: 18053.765091863515,
 30000: 16553.060333761234,
 40000: 15480.968098159508,
 50000: 13725.675568743818,
 60000: 12254.041594454073,
 70000: 10885.758024691359,
 80000: 9621.050919377652,
 90000: 8373.763416041547,
 100000: 8044.106261859582,
 125000: 6152.025366627031,
 150000: 3740.784776567541}

In [152]:
mileage_price_series=pd.Series(mileage_price)

In [153]:
mileage_price_series

5000       8750.483436
10000     20491.216000
20000     18053.765092
30000     16553.060334
40000     15480.968098
50000     13725.675569
60000     12254.041594
70000     10885.758025
80000      9621.050919
90000      8373.763416
100000     8044.106262
125000     6152.025367
150000     3740.784777
dtype: float64